robot/impl/robot_reconfigure_weak_dependencies_test.go

```diff
@@ -285,4 +291,363 @@ func TestUpdateWeakDependents(t *testing.T) {
+	// This test tests the behavior of weak dependents that have an explicit dependency.
+	// The robot has 3 components:
+	//   * base1 and base2: two fake base components.
+	//   * weak1: a weak component that depends on all components and also has an explicit dependency on base1.
+	logger := logging.NewTestLogger(t)
+
+	// Register a `Resource` that generates weak dependencies. Specifically instance of
+	// this resource will depend on every `component` resource.
+	weakAPI := resource.NewAPI(uuid.NewString(), "component", "weak")
+	weakModel := resource.DefaultModelFamily.WithModel(utils.RandomAlphaString(5))
+	weak1Name := resource.NewName(weakAPI, "weak1")
+	resource.Register(
+		weakAPI,
+		weakModel,
+		resource.Registration[*someTypeWithWeakAndStrongDeps, *someTypeWithWeakAndStrongDepsConfig]{
+			Constructor: func(
+				ctx context.Context,
+				deps resource.Dependencies,
+				conf resource.Config,
+				logger logging.Logger,
+			) (*someTypeWithWeakAndStrongDeps, error) {
+				return &someTypeWithWeakAndStrongDeps{
+					Named:     conf.ResourceName().AsNamed(),
+					resources: deps,
+				}, nil
+			},
+			WeakDependencies: []resource.Matcher{resource.TypeMatcher{Type: resource.APITypeComponentName}},
+		})
+	defer resource.Deregister(weakAPI, weakModel)
+
+	// Start robot with two components and one resource with weak dependencies
+	// (that also has an explicit dependency on one of the components).
+	// The following scenario is expected:
+	// 1) The two bases will configure first in parallel (which will each bump the logical clock).
+	// 2) The completeConfig loop will check to see if there is a need to updateWeakDependents by
+	//    checking the logical clock on also whether there are dependencies on weak dependents.
+	//    The check will fail.
+	// 3) weak1 will configure.
+	// 4) At the end of reconfiguration, weak1 will reconfigure once.
+	t.Log("Robot startup")
+	base1Name := base.Named("base1")
+	base2Name := base.Named("base2")
+	weakCfg1 := config.Config{
+		Components: []resource.Config{
+			{
+				Name:                weak1Name.Name,
+				API:                 weakAPI,
+				Model:               weakModel,
+				DependsOn:           []string{base1Name.String()},
+				ConvertedAttributes: &someTypeWithWeakAndStrongDepsConfig{},
+			},
+			{
+				Name:  base1Name.Name,
+				API:   base.API,
+				Model: fake.Model,
+			},
+			{
+				Name:  base2Name.Name,
+				API:   base.API,
+				Model: fake.Model,
+			},
+		},
+	}
+	test.That(t, weakCfg1.Ensure(false, logger), test.ShouldBeNil)
+	robot := setupLocalRobot(t, context.Background(), &weakCfg1, logger)
+
+	res, err := robot.ResourceByName(weak1Name)
+	// The resource was found and all dependencies were properly resolved.
+	test.That(t, err, test.ShouldBeNil)
+	weak1, err := resource.AsType[*someTypeWithWeakAndStrongDeps](res)
+	test.That(t, err, test.ShouldBeNil)
+	// Assert that the weak dependency was tracked.
+	test.That(t, weak1.resources, test.ShouldHaveLength, 2)
```

Comment:
> I'm guessing someTypeWithWeakAndStrongDeps saves the dependencies from res.Reconfigure(dependencies) into this local variable (named resources)?

Yep.

> And because one of the bases is a both a weak and strong dependency, that's only counted once?

Yep. My guess is only one edge in the resource graph and therefore only one dependency passed to constructor/`Reconfigure`.

Reply:
this is blowing my mind

Reply:
actually, reading the code again, I'm very wrong. We do pass in both strong and weak dependencies in both the regular reconfiguration loop and updateWeakDependents. Not sure what I got the previous idea from 😬

Reply:
Interesting discussion! Trying to close the loop on the original question here.

`weak1` will get reconfigured once with two resources (we assert those properties). One of those resources will be `base1` (also asserted on). The other must be `base2` (what else could it be?).

My question was more about why is this value not 3. The trivial response is that `Dependencies` is actually a [map](https://github.com/viamrobotics/rdk/blob/baf6c50963ec05ba848e4891deb2b938e1ba25a6/resource/resource.go?plain=1#L88) and not a slice, so if `base1` were duplicated, it'd get canceled out. But to be clear I can reason about `getDependencies(weak1)`:
* `getDependencies` first [walks the edges](https://github.com/viamrobotics/rdk/blob/baf6c50963ec05ba848e4891deb2b938e1ba25a6/robot/impl/local_robot.go?plain=1#L635-L652) and should find `base1` -- adding it to the return value
* We then [walk what comes out of](https://github.com/viamrobotics/rdk/blob/baf6c50963ec05ba848e4891deb2b938e1ba25a6/robot/impl/local_robot.go?plain=1#L654-L659) `getWeakDependencies`. Which I expect to return `base1` and `base2`. This [clause](https://github.com/viamrobotics/rdk/blob/baf6c50963ec05ba848e4891deb2b938e1ba25a6/robot/impl/local_robot.go?plain=1#L655) prevents us from re-adding `base1`, but `base2` will be added.
  * In the case where the `weakDepName` exists -- do we ever expect the `allDeps[weakDepName] != weakDepRes`?
* Leaving us with a map of base1 + base2 as the keys.